kolibri/plugins/coach/assets/src/composables/useResourceSelection.js

```diff
@@ -0,0 +1,142 @@
+  const topicId = computed(() => route.value.query.topicId);
+
+  const selectedResources = ref([]);
+
+  const bookmarksFetch = useFetch({
+    fetchMethod: () =>
+      ContentNodeResource.fetchBookmarks({
+        params: { limit: 25, available: true },
+      }),
+    fetchMoreMethod: more =>
+      ContentNodeResource.fetchBookmarks({
+        params: more,
+      }),
+    dataKey: 'results',
+    moreKey: 'more',
+    additionalDataKeys: {
+      count: 'count',
+    },
+  });
+
+  const channelsFetch = useFetch({
+    fetchMethod: () =>
+      ChannelResource.fetchCollection({
+        getParams: {
+          available: true,
+        },
+      }),
+  });
+
+  const treeFetch = useFetch({
+    fetchMethod: () =>
+      ContentNodeResource.fetchTree({ id: topicId.value, params: { include_coach_content: true } }),
+    fetchMoreMethod: more => ContentNodeResource.fetchTree({ id: topicId.value, params: more }),
+    dataKey: 'children.results',
+    moreKey: 'children.more.params',
+    additionalDataKeys: {
+      topic: '', // return the whole response as topic
+    },
+  });
+
+  const topic = computed(() => {
+    if (topicId.value) {
+      const { additionalData } = treeFetch;
+      const { topic } = additionalData.value;
+      return topic;
+    }
+    return null;
+  });
+
+  watch(topicId, () => {
+    if (topicId.value) {
+      treeFetch.fetchData();
+    }
+  });
+
+  const loading = computed(() => {
+    const sources = [bookmarksFetch, channelsFetch, treeFetch];
+
+    return sources.some(sourceFetch => sourceFetch.loading.value);
+  });
+
+  const fetchInitialData = async () => {
+    bookmarksFetch.fetchData();
+    channelsFetch.fetchData();
+    if (topicId.value) {
+      treeFetch.fetchData();
+    }
+  };
+
+  fetchInitialData();
+
+  const selectResources = (resources = []) => {
+    if (resources.length === 1) {
+      const [newResource] = resources;
+      if (!selectedResources.value.find(res => res.id === newResource.id)) {
+        selectedResources.value = [...selectedResources.value, newResource];
+      }
+    } else {
+      selectedResources.value = uniqBy([...selectedResources.value, ...resources], 'id');
+    }
+  };
+
+  const deselectResources = (resources = []) => {
+    selectedResources.value = selectedResources.value.filter(res => {
+      return !resources.find(unselectedResource => unselectedResource.id === res.id);
+    });
+  };
+
+  const setSelectedResources = (resources = []) => {
+    selectedResources.value = resources;
+  };
+
+  const selectionRules = [];
+
+  provide('topic', topic);
+  provide('channelsFetch', channelsFetch);
+  provide('bookmarksFetch', bookmarksFetch);
+  provide('treeFetch', treeFetch);
+  provide('selectionRules', selectionRules);
+  provide('selectedResources', selectedResources);
+  provide('selectResources', selectResources);
+  provide('deselectResources', deselectResources);
+  provide('setSelectedResources', setSelectedResources);
```

Comment:
I can write a version without the provide/inject, so we can weight better the implications of each pattern :)

Reply:
I just pushed a commit here in another branch removing the provide/inject: https://github.com/AlexVelezLl/kolibri/commit/5a94a593a9cf18d1db54e4c1417d83cf22748e20. Its not that bad as I first enviosioned, there are just 13 more lines of code, and although this difference will grow while we add more subPages, I'm leaning towards thinking that it is worth removing the provide/inject in favor of props passing. We can discuss more about that.

Reply:
great - thank you Alex! I'll take a look at that :)

Reply:
I've only done a quick code read through (more about overall approach of props rather than provide/inject, and not reading each line for comprehension/making sure it works) but I think this might be the way to go for now. I do think that you are right, the diff will grow, but we can figure out if and when it becomes too complex to manage. Perhaps you can stash your provide inject setup away somewhere just in case :D

Reply:
Part of the reason I'm saying this is in connection to the composable changes that we were discussing above, and weighing the pros/cons of making various changes, and which to make now vs. later. I think you lay out a good reason for using a composable for the fetching logic (although I do want to do another read and make sure I'm really understanding it) and the values of that being modular and re-usable. It's not a _huge_ refactor, and it's aligned with this work. 

For me, that seems like the "new" pattern/code to prioritize, over also introducing provide inject here. I think that approach will allow us to have the best balance of benefits of the refactoring which we really want as a core part of this project, without introducing so many changes this goes from refactor to rewrite, which I think is a different thing (and always really tempting for me.... 😸 ). What you're moving toward based on your own reflections as well as this feedback and conversation seems like a good middle ground :)